### PR TITLE
Add serverfull hosting recipe to cookbook

### PR DESCRIPTION
**Pull request by njjkgeerts**
Adding a serverfull hosting recipe to the cookbook.
Forum discussion: https://community.redwoodjs.com/t/serverfull-hosting/1182

### CODE REVIEW

**Review by Tobbe — APPROVED** [review]

**Comment by ajcwebdev** [comment]
Noticed a tiny typo in the Nginx section, essential is missing an i:

>Please note that the trailing slash in the proxy_pass value is essental to correctly map the API functions.

File: cookbook/Serverfull_Hosting.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+```terminal
+yarn deploy
+```
+
+Enjoy! 😁
+
+## Example
+
+A code example can be found at: https://github.com/njjkgeerts/redwood-pm2
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:163]: I'd recommend moving this to the top introduction section — starting with examples seems to help add meaning to the implementation steps. Also, what do you think about a link to the live example app http://redwood-pm2.nickgeerts.com ?

File: cookbook/Serverfull_Hosting.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+## Caveats
+
+Here are some caveats I have encountered:
+
+- There is a Prisma bug that occurs in Node.js version 12, so use version 14
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:169]: This is no longer the case as of Redwood v0.20.0 (including Prisma 2.9.0)

File: cookbook/Serverfull_Hosting.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+- The API only runs in fork mode in PM2, not cluster mode
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:170]: I'd recommend adding this as a callout in the applicable API setup section, e.g. in either "Project" or "PM2". You can create a callout using markdown `>` for blockquote.

> This is a blockquote example

File: cookbook/Serverfull_Hosting.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+## Requirements
+
+You should have some basic knowledge of the following tools.
+
+- Linux
+- Nginx
+- Postgres
+- PM2
+- Node
+- Yarn
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:14]: What do you think about adding links to docs/websites for these, especially Nginx and PM2?

File: cookbook/Serverfull_Hosting.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+Add Redwood's API server (in the API workspace) and PM2 (in the root) to your project.
+
+```termninal
+yarn workspace api add @redwoodjs/api-server
+yarn add -D pm2
+```
+
+Create a PM2 ecosystem configuration file. I always rename ecosystem.config.js to pm2.config.js for
+better clarity.
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:28]:
```suggestion
Create a PM2 ecosystem configuration file. For clarity, it's recommended to rename `ecosystem.config.js` to something like `pm2.config.js`.
```

Using the first person "I" can be a bit confusing. Not a real issue, but something like the above is probably better.

File: cookbook/Serverfull_Hosting.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+mv ecosystem.config.js pm2.config.js
+```
+
+Edit redwood.toml to change the API endpoint:
+
+```toml
+apiProxyPath = "/api"
+```
+
+Optionally add some scripts to your top-level package.json.
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:41]:
```suggestion
Optionally, add some scripts to your top-level package.json.
```

File: cookbook/Serverfull_Hosting.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+```json
+"scripts": {
+  "deploy:setup": "pm2 deploy pm2.config.js production setup",
+  "deploy": "pm2 deploy pm2.config.js production deploy"
+}
+```
+
+### Linux server
+
+Your server should have a user for deploying the app with. It should be configured with a SSH key pair so you can easily SSH into it from your development environment. In this example this user is named `deploy`.
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:52]:
```suggestion
Your server should have a user for deployment, which should be configured with an SSH key pair providing access to your development environment. In this example, the user is named `deploy`.
```

File: cookbook/Serverfull_Hosting.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    },
+  },
+}
+```
+
+## Deploying
+
+### Preparation
+
+First, we need to create the directories for PM2.
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:128]:
```suggestion
First, we need to create the PM2 directories.
```

File: cookbook/Serverfull_Hosting.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+```terminal
+yarn install
+yarn deploy:setup
+```
+
+Your directories at the server are now set. However, your `.env` settings are not yet configured.
+SSH into your server and create an `.env` file in the `current` subdirectory of the deploy
+directory.
```
Comment (thedavidprice) [inline cookbook/Serverfull_Hosting.md:137]:
```suggestion
Your server directories are now set. However, the `.env` settings are not yet configured.
SSH into your server and create an `.env` file in the `current` subdirectory of the deploy directory.
```

**Review by thedavidprice — CHANGES_REQUESTED** [review]
See in-line comments; mostly small grammatical and a few structure related ideas.

**Comment by thedavidprice** [comment]
Hi @njjkgeerts Thank you so much for this cookbook! It's definitely of high interest and value.

## Misc Edits
I went through the doc and suggested a lot of small edits. There were a few sentences that seemed to be split with a newline — I think I caught 'em all but do keep a lookout.

## Keywords: Nav Menu Item, Title, and Intro Paragraph
I don't have an answer to this, but I'm wondering about what keywords and descriptions we can add to help people find this Cookbook. "Serverfull" makes sense in the context of Serverless, but to the best of my knowledge, it's not a well-known term (yet). So when people are browsing the left column navigation docs, or using the site search, is there anything we can add to help others find this document and identify immediately with the topic. Some ideas:
- server hosting
- traditional or classic server architecture
- self-hosting Redwood
- hosting Redwood without serverless or docker containers

If there are some obvious ideas you want to run with, it's ok to do the following:
- have a unique Menu nave title
- and a unique Page title
- and/or include keywords and phrases in the introduction content

This isn't highest priority or blocking the PR. It's also something we can iterate on as we go. But if you have some immediate reactions and ideas, let try 'em out this round.

**Comment by njjkgeerts** [comment]
@thedavidprice can you take a look at the updates? I agree with you that "serverfull" is not a common term. "Self-hosting Redwood" would be the best of your suggestions I think.

File: cookbook/Serverfull_Hosting.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      repo,
+      path,
+      ssh_options: 'ForwardAgent=yes',
+      'post-deploy': `${build} && pm2 reload pm2.config.js --env production && pm2 save`,
+    },
+  },
+}
+```
+
+> Caveat: the API seems to only work in fork mode in PM2, not cluster mode
```
Comment (peterp) [inline cookbook/Serverfull_Hosting.md:129]: Could you explain the difference or link to the docs where they explain the difference?

**Comment by cannikin** [comment]
The opposite of “serverless” is obviously “servermore”, c’mon guys. 😁

**Comment by peterp** [comment]
I thought it was serverplusplus

**Comment by thedavidprice** [comment]
@njjkgeerts all the changes look great to me! Review approved on my end. Thanks 🚀

@peterp I'll let you take it from here.

**Review by thedavidprice — APPROVED** [review]

File: lib/build.js [hunk 11]
```diff
@@ -130,6 +130,7 @@ const SECTIONS = [
       { pageBreakAtHeadingDepth: [1], file: './cookbook/GoTrue_Auth.md' },
       { pageBreakAtHeadingDepth: [1], file: './cookbook/Role-based_Access_Control.md' },
       { pageBreakAtHeadingDepth: [1], file: './cookbook/Pagination.md' },
+      { pageBreakAtHeadingDepth: [1], file: './cookbook/Serverfull_Hosting.md' },
```
Comment (thedavidprice) [inline lib/build.js:133]:
```suggestion
      { pageBreakAtHeadingDepth: [1], 
      file: './cookbook/Serverfull_Hosting.md',
      title: 'Self-hosting Redwood' },
```

**Comment by thedavidprice** [comment]
Note: changed title of menu item. Could be just "Self-hosting". Open to suggestions otherwise.

Screen Shot 2020-10-29 at 9 18 56 PM

File: cookbook/Serverfull_Hosting.md (hunk 10 above)
Comment (njjkgeerts) [reply cookbook/Serverfull_Hosting.md]: I also added the link in the document.
https://pm2.keymetrics.io/docs/usage/cluster-mode/

File: cookbook/Self-hosting_Redwood.md [hunk 12]
```diff
@@ -1,6 +1,6 @@
-# Serverfull Hosting
+# Self-hosting Redwood
```
Comment (thedavidprice) [inline cookbook/Self-hosting_Redwood.md:1]:
```suggestion
# Self-hosting Redwood: Serverfull
```

For a term where it's unclear what the user is searching for, it's ok to have the Menu Title and the Page title be different (as long as they're intuitively related). The site search can pick up more keywords that way.

In this case, not sure if we _need_ to make the change I've suggested here, but I'm fine if we do. Or if there's a better title, please suggest!

And thank you for taking the time to think about keywords with me. It's important! If people can't find an article, then what's the point? 😉

**Comment by thedavidprice** [comment]
All systems go! Thanks again @njjkgeerts 🚀

**Pull request merged**